compilers/concrete-compiler/compiler/lib/Dialect/FHE/IR/FHEOps.cpp

```diff
@@ -394,6 +394,158 @@ void MulEintIntOp::getCanonicalizationPatterns(
+        return mlir::failure();
+      }
+
+      auto intermediateTableContent =
+          (intermediateTableContentAttr.getValues<int64_t>());
+      auto currentTableContent = (currentTableContentAttr.getValues<int64_t>());
+
+      auto inputValue = intermediateOperation.getA();
+      auto inputType = inputValue.getType().dyn_cast<FheIntegerInterface>();
+      auto inputBitWidth = inputType.getWidth();
+      auto inputIsSigned = inputType.isSigned();
+
+      auto intermediateType =
+          (intermediateValue.getType().dyn_cast<FheIntegerInterface>());
+      auto intermediateBitWidth = intermediateType.getWidth();
+      auto intermediateIsSigned = intermediateType.isSigned();
+
+      auto usersOfPreviousOperation = intermediateOperation->getUsers();
+      auto numberOfUsersOfPreviousOperation = std::distance(
+          usersOfPreviousOperation.begin(), usersOfPreviousOperation.end());
+
+      if (numberOfUsersOfPreviousOperation > 1) {
+        // This is a special case.
+        //
+        // Imagine you have this structure:
+        // -----------------
+        // x: uint6
+        // y: uint3 = tlu[x]
+        // z: uint3 = y + 1
+        // a: uint3 = tlu[y]
+        // b: uint3 = a + z
+        // -----------------
+        //
+        // In this case, it's be better not to fuse `a = tlu[tlu[x]]`.
+        //
+        // The reason is that intermediate `y` is necessary for `z`,
+        // so it has to be computed anyway.
+        //
+        // For calculate `a` there are 2 options:
+        // - fused tlu on x
+        // - regular tlu on y
+        //
+        // So for such cases, it's only better to fuse if the
+        // bit width of `x` is smaller than the bit width of `y`.
+
+        auto shouldFuse = inputBitWidth < intermediateBitWidth;
+        if (!shouldFuse) {
+          return mlir::failure();
+        }
+      }
+
+      auto intermediateTableSize = 1 << inputBitWidth;
+      auto currentTableSize = 1 << intermediateBitWidth;
+
+      auto newTableContent = std::vector<int64_t>();
+      newTableContent.reserve(intermediateTableSize);
+
+      auto lookup = [&](ssize_t index) {
+        if (index < 0) {
+          index += intermediateTableSize;
+        }
+        auto resultOfFirstLookup = intermediateTableContent[index];
+
+        if (resultOfFirstLookup < 0) {
+          resultOfFirstLookup += currentTableSize;
+          if (resultOfFirstLookup < 0) {
+            resultOfFirstLookup = currentTableSize / 2;
+          }
+        } else if (resultOfFirstLookup >= currentTableSize) {
+          if (!intermediateIsSigned) {
+            resultOfFirstLookup = currentTableSize - 1;
+          } else {
+            resultOfFirstLookup = (currentTableSize / 2) - 1;
+          }
+        }
+        auto resultOfSecondLookup = currentTableContent[resultOfFirstLookup];
+
+        return resultOfSecondLookup;
+      };
+
+      if (!inputIsSigned) {
```

Comment:
Same here. Add a comment why signed and unsigned values need to be treated differently.